sodium/src/aead.rs

```diff
@@ -0,0 +1,242 @@
+//! This module provides access to libsodium
+
+use super::{check_init, secbuf::SecBuf};
+use crate::error::SodiumResult;
+
+/// Used to set the size of nonce var in the enc fns
+pub const NONCEBYTES: usize =
+    rust_sodium_sys::crypto_aead_xchacha20poly1305_ietf_NPUBBYTES as usize;
+/// Used with the size of the message to set the size of the cipher and decripted message
+///
+/// Note: look at the test cases to see how it is used
+pub const ABYTES: usize = rust_sodium_sys::crypto_aead_xchacha20poly1305_ietf_ABYTES as usize;
+
+/// Generate symmetric cipher text given a message, secret, and optional auth data
+///
+/// @param {SecBuf} message - data to encrypt
+///
+/// @param {SecBuf} secret - symmetric secret key
+///
+/// @param {SecBuf} adata - optional additional authenticated data
+///
+/// @param {SecBuf} nonce - Empty Buffer to be used as output
```

Comment:
This is actually input. It is pub data that must be sent along with the message (or derived through some sequence scheme) in order to decrypt it.

Reply:
Yes, this is sent back because it is needed while decrypting. The statement should have said that the implementor needs to pass in an empty buffer in which the output value will be stored.
Am I getting that right?

Reply:
no, this is a read param, you can see it does a read_lock() and passes into the ffi encrypt immutable. As you mention in the decrypt, it is the initialization vector, but it is still user specified even in the encrypt.